src/nf/nf_reshape2d_layer_submodule.f90

```diff
@@ -0,0 +1,50 @@
+submodule(nf_reshape2d_layer) nf_reshape2d_layer_submodule
+
+  use nf_base_layer, only: base_layer
+
+  implicit none
+
+contains
+
+  pure module function reshape2d_layer_cons(output_shape) result(res)
+    integer, intent(in) :: output_shape(2)
+    type(reshape2d_layer) :: res
+    res % output_shape = output_shape
+  end function reshape2d_layer_cons
+
+
+  pure module subroutine backward(self, input, gradient)
+    class(reshape2d_layer), intent(in out) :: self
+    real, intent(in) :: input(:)
+    real, intent(in) :: gradient(:,:)
+    ! The `input` dummy argument is not used but nevertheless declared
+    ! because the abstract type requires it.
+    self % gradient = pack(gradient, .true.)
+  end subroutine backward
+
+
+  pure module subroutine forward(self, input)
+    class(reshape2d_layer), intent(in out) :: self
+    real, intent(in) :: input(:)
+    self % output = reshape(input, self % output_shape)
```

Comment:
A pointer could be used here, instead of `reshape` to avoid a tmp copy of `input`